File: src/test/kotlin/io/github/muindor/tcresearchsolver/solver/CostTest.kt
package io.github.muindor.tcresearchsolver.solver

import org.junit.jupiter.api.Assertions.*
import org.junit.jupiter.api.Test

class CostTest {

    @Test
    fun `orders by scarcity first, then cells`() {
        assertTrue(compareCost(Cost(1.0, 100.0), Cost(2.0, 0.0)) < 0)
        assertTrue(compareCost(Cost(2.0, 1.0), Cost(2.0, 3.0)) < 0)
        assertEquals(0, compareCost(Cost(2.0, 3.0), Cost(2.0, 3.0)))
    }

    @Test
    fun `adds componentwise (Infinity-safe)`() {
        assertEquals(Cost(4.0, 6.0), addCost(Cost(1.0, 2.0), Cost(3.0, 4.0)))
        assertEquals(Double.POSITIVE_INFINITY, addCost(INF_COST, ZERO_COST).scarcity)
    }

    @Test
    fun `lessThan is strict`() {
        assertTrue(lessThan(Cost(1.0, 0.0), Cost(1.0, 1.0)))
        assertFalse(lessThan(Cost(1.0, 1.0), Cost(1.0, 1.0)))
    }

    @Test
    fun `INF_COST compares greater than any finite cost`() {
        assertTrue(compareCost(INF_COST, Cost(1e9, 1e9)) > 0)
        assertTrue(lessThan(Cost(1.0, 1.0), INF_COST))
        assertEquals(0, compareCost(INF_COST, INF_COST))
    }
}

File: reference/ts-solver/tests/core/cost.test.ts
import { describe, expect, it } from 'bun:test';
import { type Cost, compareCost, addCost, ZERO_COST, INF_COST, lessThan } from '../../app/src/core/cost';

describe('lexicographic cost (scarcity, cells)', () => {
  it('orders by scarcity first, then cells', () => {
    expect(compareCost({ scarcity: 1, cells: 100 }, { scarcity: 2, cells: 0 })).toBeLessThan(0);
    expect(compareCost({ scarcity: 2, cells: 1 }, { scarcity: 2, cells: 3 })).toBeLessThan(0);
    expect(compareCost({ scarcity: 2, cells: 3 }, { scarcity: 2, cells: 3 })).toBe(0);
  });
  it('adds componentwise (Infinity-safe)', () => {
    expect(addCost({ scarcity: 1, cells: 2 }, { scarcity: 3, cells: 4 })).toEqual({ scarcity: 4, cells: 6 });
    expect(addCost(INF_COST, ZERO_COST).scarcity).toBe(Number.POSITIVE_INFINITY);
  });
  it('lessThan is strict', () => {
    expect(lessThan({ scarcity: 1, cells: 0 }, { scarcity: 1, cells: 1 })).toBe(true);
    expect(lessThan({ scarcity: 1, cells: 1 }, { scarcity: 1, cells: 1 })).toBe(false);
  });
});
